api/envoy/service/accesslog/v3/als.proto

```diff
@@ -27,6 +27,10 @@ service AccessLogService {
   // expectation that it might be lossy.
   rpc StreamAccessLogs(stream StreamAccessLogsMessage) returns (StreamAccessLogsResponse) {
   }
+
+  rpc BufferedCriticalAccessLogs(StreamAccessLogsMessage)
```

Comment:
This isn't bidi streaming but unary, is this intentional?

Reply:
I'm considering that we don't have to apply bidi streaming here. This is because 

1. If we apply bidi streaming here, we should setup timer per message to measure ACK/NACK waiting time. In this situation, we can use stream timeout directly when we use unary call.
2. If we apply bidi streaming here, we should have some indicator per message to track status (e.g. UUID) that will be used whether to determine buffer or not by the ACK/NACK response. But with unary call, we don't have to do it because single stream would bind to single message. 

Stream establishment pressure won't be reduced because this periodical operation won't occur frequently. cc @snowp 

But, +1 to have we have buffering async client. Do you have any idea about this?

Reply:
A few things to consider in favor of a bidi stream:
- we currently issue the `Identifier` only once per stream, which amortizes the cost of this payload (which is non-trivial as it contains all registered extensions and potentially a large metadata) over all the requests within that stream. Using unary we lose this, which means that if we were to send small logging payloads, the majority of the request might actually be the identifier.
- how would you deal with the case when we want to flush but there is already an outstanding request? You could withhold the logs until it is no longer outstanding, though this would reduce throughput (you lose out on something akin to pipelining) somewhat

>Stream establishment pressure will be reduced because this periodical operation won't occur frequently. 
Not sure what you mean here, are you saying that we promise that we won't issue these logs often so it's not a problem?  It might be good to design a system that doesn't rely on the access pattern, it seems a bit brittle (e.g. we only do this on certain errors, but some incident causes this to happen a lot, so we suffer cascading failures)

Reply:
As for a generic client, I think for unary the existing retry functionality somewhat captures a generic caching client, so you'd want to follow the same pattern (or extend the Router to support this) in order to get a client that can do the appropriate buffering/payload merger/whatever you need.

For a bidi stream you could use a client that works with an assumes an unordered 1-1 request/response pattern, where each request/response carries an ID that allows for correlation, with a few virtual functions that lets you decide when to buffer a request and optionally how to combine multiple buffered requests into one. For example:

```
template<class RequestT, class ResponseT>
class BufferedAsyncClient : AsyncClient<RequestT, ResponseT> {
  virtual uint64_t requestId(const RequestT& req) PURE;
  virtual uint64_t responseId(const ResponseT& res) PURE;
  virtual bool accepted(const RequestT& req, const ResponseT& res) PURE;
  virtual std::vector<std::unqiue_ptr<RequestT>> prepareRequests(std::vector<std::unique_ptr<RequestT>>&& requests) PURE;
};
```

This would allow for a fairly generic handling of issuing request, tracking the outstanding responses, evaluating the response and its correlated request, and optionally combining multiple buffered requests into one using whatever policy
fits the use case.

With this in place, it would be fairly straightforward to add new features, like a max time to buffer requests for, max number of requests to buffer, etc., and making this generally available to all systems that might want to use it.

Reply:
> we currently issue the Identifier only once per stream, which amortizes the cost of this payload (which is non-trivial as it contains all registered extensions and potentially a large metadata) over all the requests within that stream. Using unary we lose this, which means that if we were to send small logging payloads, the majority of the request might actually be the identifier.

I agree with this Idea. It may be very redundant.

> how would you deal with the case when we want to flush but there is already an outstanding request? You could withhold the logs until it is no longer outstanding, though this would reduce throughput (you lose out on something akin to pipelining) somewhat

This is also the case in bidi not only unary, although the message needs to be stalled until the Unary stream is complete. The steps I consider are as follows.

For each message stored in the buffer, store an ID to identify the message and metadata about the processing state. The processing state is defined as both Initial and Pending, and when flushing, the buffer status becomes Pending, and if an ACK is received, it is removed from the buffer; if a NACK is received, or a response timeout occurs, it becomes Pending -> Initial, and the buffer becomes stagnant. Initial if a NACK is received or the response times out, and continues to dwell in the buffer.

This process may overwhelm the buffer and degrade performance, but it does not compromise pipeline-like processing. This is because everything is accomplished by an asynchronous process that tracks the state.

Reply:
As for a generic client, your proposal seems very cool to me. let me to refactor my code.

Reply:
I think one thing we should make clear in documentation and/or an issue is the various use cases that this makes sense. It would help in understanding some of the tradeoffs. When is an ALS endpoint expected to NACK? When do we expect stream timeouts? Why is the ACK/NACK wait time important?

I'm sure these all have very good answers (and I can imagine what some of them are). But best to be explicit.

Reply:
Sure. I'm also considering about describing tradeoffs explicitly in docs because this feature potentially decrease perf.